### PR TITLE
CSHARP-4708: Support expressions created by Expression.MakeIndex.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

**Comment by rstam** [comment]
I've rebased on master and pushed a new commit. This work is now ready to review.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/IDictionaryMethod.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+*/
+
+using System.Reflection;
+using MongoDB.Driver.Linq.Linq3Implementation.Misc;
+
+namespace MongoDB.Driver.Linq.Linq3Implementation.Reflection
+{
+    internal static class IDictionaryMethod
+    {
+        // public methods
```
Comment (sanych-sun) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/IDictionaryMethod.cs:23]: Minor: public static methods

File: src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/IListMethod.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+*/
+
+using System.Reflection;
+using MongoDB.Driver.Linq.Linq3Implementation.Misc;
+
+namespace MongoDB.Driver.Linq.Linq3Implementation.Reflection
+{
+    internal static class IListMethod
+    {
+        // public methods
```
Comment (sanych-sun) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/IListMethod.cs:23]: Minor: public static methods

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4708Tests.cs [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+                new C { Id = 2, Dictionary = new Dictionary<string, int> { ["b"] = 2 }, Document = new BsonDocument("b", 2), List = new List<int> { 2 } });
+            return collection;
+        }
+
+        private class C
+        {
+            public int Id { get; set; }
+            public IDictionary<string, int> Dictionary { get; set; }
+            public BsonDocument Document { get; set; }
+            public IList<int> List { get; set; }
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4708Tests.cs:695]: Should we test accessing to array elements here or this is out of scope for the current PR?

**Review by sanych-sun — APPROVED** [review]
LGTM + minor comments

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Review by sanych-sun — APPROVED** [review]
LGTM
